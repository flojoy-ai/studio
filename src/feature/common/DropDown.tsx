import { Box, createStyles } from "@mantine/core";
import React, { useRef } from "react";
const useStyles = createStyles((theme) => {
  return {
    dropdownWrapper: {
      position: "relative",
    },
    dropdownContainer: {
      paddingTop: 15,
      paddingBottom: 15,
      paddingLeft: 7,
      paddingRight: 7,
      borderRadius: 15,
      transition: "all ease 0.5s",
      flexDirection: "column",
      position: "absolute",
      top: "52px",
      right: 0,
      width: "fit-content",
      alignItems: "flex-start",
      zIndex: -1,
      transform: "translateY(-10%)",
      opacity: 0,
      backgroundColor: theme.colors.modal[1],
      boxShadow: theme.colorScheme === "light" ? theme.shadows.sm : "none",
      "> button": {
        padding: "8px 12px",
        marginRight: "10px",
        cursor: "pointer",
        borderRadius: 2,
        fontSize: "14px",
        fontWeight: "bold",
        textDecoration: "none",
        background: theme.colors.modal[1],
        border: "none",
        width: "100%",
        textAlign: "start",
        whiteSpace: "nowrap",
        color: theme.colors.title[0],
      },
      "> button:not(.disabled):hover": {
        backgroundColor:
          theme.colorScheme === "dark"
            ? theme.colors.dark[5]
            : theme.colors.modal[0],
      },
    },
  };
});
interface DropDownProps {
  children: React.ReactNode;
  dropDownBtn: React.ReactNode;
}
const DropDown = ({ children, dropDownBtn }: DropDownProps) => {
  const { classes } = useStyles();
  const DropDownElem = useRef<HTMLDivElement>(null);
  const openDropDown = () => {
    if (DropDownElem.current) {
      DropDownElem.current.style.opacity = "1";
      DropDownElem.current.style.zIndex = "50";
      DropDownElem.current.style.transform = "translateY(0)";
    }
  };
  const closeDropDown = () => {
    if (DropDownElem.current) {
      DropDownElem.current.style.opacity = "0";
      DropDownElem.current.style.zIndex = "-1";
      DropDownElem.current.style.transform = "translateY(-10%)";
    }
  };
  return (
    <Box
      data-cy="dropdown-wrapper"
      data-testid="dropdown-wrapper"
      className={classes.dropdownWrapper}
      onMouseEnter={openDropDown}
      onMouseLeave={closeDropDown}
    >
      {dropDownBtn}
      <Box
        data-testid="dropdown-container"
        className={classes.dropdownContainer}
        ref={DropDownElem}
        onMouseLeave={closeDropDown}
      >
        {children}
      </Box>
    </Box>
  );
};
export default DropDown;
