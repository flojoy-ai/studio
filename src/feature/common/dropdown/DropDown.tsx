import React, { useRef } from "react";
import "@src/feature/common/dropdown/dropdown.css";

interface DropDownProps {
  children: React.ReactNode;
  DropDownBtn: React.ReactNode;
  theme: "light" | "dark";
}
const DropDown = ({ children, DropDownBtn, theme }: DropDownProps) => {
  const DropDownElem = useRef<HTMLDivElement>(null);
  const openDropDown = () => {
    if (DropDownElem.current) {
      DropDownElem.current.style.opacity = "1";
      DropDownElem.current.style.zIndex = "50";
      DropDownElem.current.style.transform = "translateY(0)";
    }
  };
  const closeDropDown = () => {
    if (DropDownElem.current) {
      DropDownElem.current.style.opacity = "0";
      DropDownElem.current.style.zIndex = "-1";
      DropDownElem.current.style.transform = "translateY(-10%)";
    }
  };
  return (
    <div
      data-testid="dropdown-wrapper"
      className="dropdown__wrapper"
      onMouseEnter={openDropDown}
      onMouseLeave={closeDropDown}
    >
      {DropDownBtn}
      <div
        data-testid="dropdown-container"
        className={`dropdown__container ${theme}`}
        ref={DropDownElem}
        onMouseLeave={closeDropDown}
      >
        {children}
      </div>
    </div>
  );
};

export default DropDown;
