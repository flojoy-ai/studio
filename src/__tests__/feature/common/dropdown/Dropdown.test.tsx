import "@testing-library/jest-dom/extend-expect"
import "@testing-library/jest-dom"
import { render, fireEvent } from "@testing-library/react";
import DropDown from "@src/feature/common/dropdown/DropDown"; 

const theme = "light";
describe("DropDown", () => {
  const DropDownBtn = <button>Dropdown Button</button>;
  const children = <div>Dropdown Children</div>;
  it("should render correctly", () => {
    const { container, getByText, getByTestId } = render(
      <DropDown DropDownBtn={DropDownBtn} theme={theme}>
        {children}
      </DropDown>
    );
    // Check that the button is rendered
    const button = getByText("Dropdown Button");
    expect(button).toBeInTheDocument();

    // Check that the dropdown content is not visible
    const dropdownContent = getByTestId("dropdown-container");
    expect(dropdownContent).toHaveClass("dropdown__container", theme)
    expect(container).toMatchSnapshot();
  });

    it("should open and close on hover", () => {
      const { getByTestId } = render(
        <DropDown DropDownBtn={DropDownBtn} theme={theme}>
          {children}
        </DropDown>
      );
      const dropdownWrapper = getByTestId("dropdown-wrapper");
      const dropdownContainer = getByTestId("dropdown-container");

      // Check that dropdown is closed initially
      expect(dropdownContainer).toHaveClass("dropdown__container");
      
      // Open dropdown on mouse enter
      fireEvent.mouseEnter(dropdownWrapper);
      expect(dropdownContainer).toHaveStyle("opacity: 1");
      expect(dropdownContainer).toHaveStyle("z-index: 50");
      expect(dropdownContainer).toHaveStyle("transform: translateY(0)");

      // Close dropdown on mouse leave
      fireEvent.mouseLeave(dropdownWrapper);
      expect(dropdownContainer).toHaveStyle("opacity: 0");
      expect(dropdownContainer).toHaveStyle("z-index: -1");
      expect(dropdownContainer).toHaveStyle("transform: translateY(-10%)");
    });
});
