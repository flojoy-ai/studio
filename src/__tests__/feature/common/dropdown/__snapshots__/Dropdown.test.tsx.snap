// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown should render correctly 1`] = `
<div>
  <div
    class="dropdown__wrapper"
    data-cy="dropdown-wrapper"
    data-testid="dropdown-wrapper"
  >
    <button>
      Dropdown Button
    </button>
    <div
      class="dropdown__container light"
      data-testid="dropdown-container"
    >
      <div>
        Dropdown Children
      </div>
    </div>
  </div>
</div>
`;
